engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java

```diff
@@ -128,27 +161,227 @@ public BulletPhysics() {
+                    CollisionGroup.combineGroups(body.collisionGroup),
+                    CollisionGroup.combineGroups(body.collidesWith));
+        } else {
+            newRigidBody(null, entity, body, location);
+        }
     }
-    //*****************Physics Interface methods******************\\
 
-    @Override
-    public List<PhysicsSystem.CollisionPair> getCollisionPairs() {
-        List<PhysicsSystem.CollisionPair> temp = collisions;
-        collisions = new ArrayList<>();
-        return temp;
+    @ReceiveEvent(components = {RigidBodyComponent.class, LocationComponent.class})
+    public void removeRigidBody(BeforeDeactivateComponent event, EntityRef entity) {
+        entityRigidBodies.computeIfPresent(entity, (entityRef, rb) -> {
+            Matrix4f m = new Matrix4f();
+            Vector3f aabbMin = new Vector3f();
+            Vector3f aabbMax = new Vector3f();
+            rb.getCollisionShape().getAabb(m, aabbMin, aabbMax);
+            awakenArea(rb.getWorldTransform().getTranslation(new Vector3f()), (aabbMax.sub(aabbMin)).length() * .5f);
+
+            discreteDynamicsWorld.removeRigidBody(rb);
+            rb.dispose();
+            return null;
+        });
+    }
+
+    @ReceiveEvent(components = RigidBodyComponent.class)
+    public void onImpulse(ImpulseEvent event, EntityRef entity) {
+        btRigidBody bt = entityRigidBodies.get(entity);
+        if (bt != null) {
+            bt.applyCentralImpulse(event.getImpulse());
+        }
+    }
+
+    @ReceiveEvent(components = RigidBodyComponent.class)
+    public void onForce(ForceEvent event, EntityRef entity) {
+        btRigidBody bt = entityRigidBodies.get(entity);
+        if (bt != null) {
+            bt.applyCentralForce(event.getForce());
+        }
+    }
+
+    @ReceiveEvent(components = RigidBodyComponent.class)
+    public void onChangeVelocity(ChangeVelocityEvent event, EntityRef entity) {
+        btRigidBody bt = entityRigidBodies.get(entity);
+        if (bt != null) {
+            if (event.getAngularVelocity() != null) {
+                bt.setAngularVelocity(event.getAngularVelocity());
+            }
+            if (event.getLinearVelocity() != null) {
+                bt.setLinearVelocity(event.getLinearVelocity());
+            }
+        }
+    }
+
+    @ReceiveEvent(components = BlockComponent.class)
+    public void onBlockAltered(OnChangedBlock event, EntityRef entity) {
+        awakenArea(new Vector3f(event.getBlockPosition()), 0.6f);
+    }
+
+
+    @ReceiveEvent(components = {LocationComponent.class, RigidBodyComponent.class})
+    public void onItemImpact(ImpactEvent event, EntityRef entity) {
+        btRigidBody rb = entityRigidBodies.get(entity);
+        if (rb != null) {
+            Vector3f vImpactNormal = new Vector3f(event.getImpactNormal());
+            Vector3f vImpactPoint = new Vector3f(event.getImpactPoint());
+            Vector3f vImpactSpeed = new Vector3f(event.getImpactSpeed());
```

Comment:
This allocates new vectors on each call of this event handler (similar cases in the other handlers). I would imagine that this happens quite often, and could be a good candidate for optimizations to reuse instances.

I do appreciate that we have readable code in the first place, so maybe we can just add a TODO comment in this file to look at this later and eventually address it in a follow up PR.

Reply:
umm, maybe something to consider through profiling.